docs: Fuel 101 - Relationship to Ethereum (#216)
FuelLabs/docs-hub: docs/fuel-101/relationship-to-ethereum.mdx

### Enhanced Throughput

The parallel processing capability of the FuelVM directly translates into enhanced transaction throughput. By leveraging the computational resources available more effectively, the FuelVM can handle a significantly higher volume of transactions compared to other Ethereum-aligned architectures. This increase in throughput is crucial for supporting the growing number of dApps, app chains and L3s on Ethereum, many of which require high transaction capacities to function effectively​. 

### Optimized CPU Utilization

The FuelVM introduces a paradigm shift in the utilization of computational resources within blockchain networks, particularly addressing the inefficiencies prevalent in other systems. By implementing strategies that allow for more extensive use of a CPU's threads and cores, the FuelVM enhances the Ethereum network's performance and scalability.

> At the core of FuelVM's approach to optimized CPU utilization is its use of the Unspent Transaction Output (UTXO) model, complemented by strict state access lists. This allows the FuelVM to precisely determine which parts of the state will be read or written by each transaction before its execution. This topic will be covered in more detail in the next section on solving for state bloat in Ethereum.

Reviewer: nit: just how we handle the FuelVM and EVM acronyms we should move "Unspent Transaction Output (UTXO)" to the first time we mention it and then refer to it as UTXO throughout the rest of the document

```suggestion
At the core of FuelVM's approach to optimized CPU utilization is its use of the Unspent Transaction Output (UTXO) model, complemented by strict state access lists. This allows the FuelVM to precisely determine which parts of the state will be read or written by each transaction before its execution. This topic will be covered in more detail in the next section on solving for state bloat in Ethereum.
```

Fuel's approach to asset management, in which all asset transfers affect only a single state element, streamlines transactions. Native assets are managed more efficiently, without the need for complex state changes.

### Eliminating Approval State

One of the innovative aspects of Fuel's state approach is the elimination of the approval state. In traditional Ethereum transactions, the `approve` and `transferFrom` functions involve multiple state changes that contribute to state bloat and increased transaction costs. Fuel's model does away with this by simplifying the asset transfer process, thereby enhancing transaction efficiency and reducing unnecessary state modifications.

### State Minimized Techniques

Fuel introduces several state-minimized techniques that leverage bandwidth and execution over state storage, presenting developers with new ways for building other than just smart contract storage:

* Scripts: Ephemeral logic is included in transactions, not stored in state. Unlike EVM transactions which can call a contract directly (but can only call a single contract), Fuel transactions execute a script, which may call zero or more contracts.

> * Predicates: Lightweight, stateless contracts. A predicate is a new, pure mechanism for authorizing transactions. A predicate can only access the data in a transaction, it cannot view the current chain state. Predicates can be used, among other things, to enable native (stateless) account abstraction.

Reviewer: IMO I would avoid the word contract when discussing predicates.  Mostly because we don't have to deploy predicates, but have to deploy contracts.

```suggestion
* Predicates: Lightweight, stateless programs. A predicate is a new, pure mechanism for authorizing transactions. A predicate can only access the data in a transaction, it cannot view the current chain state. Predicates can be used, among other things, to enable native (stateless) account abstraction.
```

---
title: Relationship to Ethereum
category: Fuel 101
parent:
  label: Fuel 101
  link: /fuel-101
---

# Relationship to Ethereum

Fuel is an operating system purpose built for Ethereum Rollups. It allows for rollups to solve for parallelization, state minimized execution and interoperability in Ethereum. Here is how Fuel fits into the Ethereum ecosystem and improves the scalability of the network:

- **Fuel brings long needed improvements to the Ethereum ecosystem.** Due to constraints with backward compatability, the Ethereum virtual machine (EVM) has been limited in its ability to scale. The Fuel virtual machine (FuelVM) brings those improvements to Ethereum.

> - **Fuel reduces Ethereum state bloat.** Fuel utilizes the UTXO model for strict state access lists, which allows for a more efficient usage of state storage on chain.

Reviewer 2: ```suggestion
- **Fuel reduces Ethereum state bloat.** Fuel utilizes the UTXO model for [strict state access lists](https://github.com/ethereum/EIPs/issues/648), which allows for a more efficient usage of state storage on chain.
```

Reviewer 2: A quick link to the Ethereum EIP for people who don't know what strict access lists are

- **Interoperability with Ethereum built in.** Compatibility with Ethereum is essential to Fuel, and being able to build code, send transactions and interoperate across the networks is incorporated into every aspect of Fuel's design.

Let's look at each of these areas in detail.

## The FuelVM and Improving Ethereum Scalability

The introduction of the Fuel Virtual Machine (FuelVM) into the Ethereum ecosystem represents a significant step forward in addressing the scalability challenges that have historically constrained Ethereum's growth and performance.
The FuelVM introduces several key innovations designed to improve transaction throughput, reduce latency, and optimize resource utilization on the Ethereum network

### Parallel Transaction Execution

> One of the most significant features of the FuelVM is its ability to execute transactions in parallel. This is achieved through the use of strict state access lists, which enable the FuelVM to identify dependencies between transactions ahead of their execution. By mapping out these dependencies, the FuelVM can efficiently allocate transactions to multiple threads or cores of a CPU, drastically increasing the number of transactions processed within a given time frame. This approach contrasts with the Ethereum Virtual Machine (EVM), which primarily processes transactions sequentially, leading to potential bottlenecks and increased latency during periods of high network demand.

Reviewer 2: ```suggestion
One of the most significant features of the FuelVM is its ability to execute transactions in parallel. This is achieved through the use of [strict state access lists](https://github.com/ethereum/EIPs/issues/648), which enable the FuelVM to identify dependencies between transactions ahead of their execution. By mapping out these dependencies, the FuelVM can efficiently allocate transactions to multiple threads or cores of a CPU, drastically increasing the number of transactions processed within a given time frame. This approach contrasts with the Ethereum Virtual Machine (EVM), which primarily processes transactions sequentially, leading to potential bottlenecks and increased latency during periods of high network demand.
```

### Enhanced Throughput

The parallel processing capability of the FuelVM directly translates into enhanced transaction throughput. By leveraging the computational resources available more effectively, the FuelVM can handle a significantly higher volume of transactions compared to other Ethereum-aligned architectures. This increase in throughput is crucial for supporting the growing number of dApps, app chains and L3s on Ethereum, many of which require high transaction capacities to function effectively​.

### Optimized CPU Utilization

The FuelVM introduces a paradigm shift in the utilization of computational resources within blockchain networks, particularly addressing the inefficiencies prevalent in other systems. By implementing strategies that allow for more extensive use of a CPU's threads and cores, the FuelVM enhances the Ethereum network's performance and scalability.

> At the core of FuelVM's approach to optimized CPU utilization is its use of the Unspent Transaction Output (UTXO) model, complemented by strict state access lists. This allows the FuelVM to precisely determine which parts of the state will be read or written by each transaction before its execution. This topic will be covered in more detail in the next section on solving for state bloat in Ethereum.

Reviewer 2: ```suggestion
At the core of FuelVM's approach to optimized CPU utilization is its use of the Unspent Transaction Output (UTXO) model, complemented by [strict state access lists](https://github.com/ethereum/EIPs/issues/648). This allows the FuelVM to precisely determine which parts of the state will be read or written by each transaction before its execution. This topic will be covered in more detail in the next section on solving for state bloat in Ethereum.
```

## Solving for Ethereum State Bloat

Fuel's architectural design introduces a new approach to managing and mitigating the challenges associated with Ethereum's state bloat, a constant issue that has implications for the network's scalability, efficiency, and overall performance. At the root of Fuel's strategy is a transaction model and state management system designed to optimize resource usage while maintaining the network's integrity and security.

### Localized State Trees

Fuel's use of the UTXO model changes how state is managed and stored on chain. Unlike the global state tree utilized by Ethereum, which requires updating the entire tree for each transaction, Fuel adopts localized state trees. This means that each smart contract has its own state tree, significantly reducing the overhead associated with state updates.

Fuel's approach to asset management, in which all asset transfers affect only a single state element, streamlines transactions. Native assets are managed more efficiently, without the need for complex state changes.

### Eliminating Approval State

> One of the innovative aspects of Fuel's state approach is the elimination of the approval state. In traditional Ethereum transactions, the `approve` and `transferFrom` functions involve multiple state changes that contribute to state bloat and increased transaction costs. Fuel's model does away with this by simplifying the asset transfer process, thereby enhancing transaction efficiency and reducing unnecessary state modifications.

Reviewer 2: ```suggestion
One of the innovative aspects of Fuel's state approach is the elimination of the approval state. In traditional Ethereum transactions, the `approve` and `transferFrom` functions involve multiple state changes that contribute to state bloat and increased transaction costs. Fuel's model streamlines the process by treating all assets natively. This means that although you still require a smart contract for minting and burning fungible assets, the sending and receiving of these assets can be conducted independently of the asset contract. By treating all assets as first-class citizens, it significantly enhances transaction efficiency and reduces unnecessary state modifications.
```

### State Minimized Techniques

Fuel introduces several state-minimized techniques that leverage bandwidth and execution over state storage, presenting developers with new ways for building other than just smart contract storage:

> - Scripts: Ephemeral logic is included in transactions, not stored in state. Unlike EVM transactions which can call a contract directly (but can only call a single contract), Fuel transactions execute a script, which may call zero or more contracts.

Reviewer 2: ```suggestion
- [Scripts](https://docs.fuel.network/docs/sway/sway-program-types/scripts/): Ephemeral logic is included in transactions, not stored in state. Unlike EVM transactions which can call a contract directly (but can only call a single contract), Fuel transactions execute a script, which may call zero or more contracts.
```

> - Predicates: Lightweight, stateless programs. A predicate is a new, pure mechanism for authorizing transactions. A predicate can only access the data in a transaction, it cannot view the current chain state. Predicates can be used, among other things, to enable native (stateless) account abstraction.

Reviewer 2: ```suggestion
- [Predicates](https://docs.fuel.network/guides/intro-to-predicates/): Lightweight, stateless programs. A predicate is a new, pure mechanism for authorizing transactions. A predicate can only access the data in a transaction, it cannot view the current chain state. Predicates can be used, among other things, to enable native (stateless) account abstraction.
```

Reviewer 2: The sway repo's program type reference is not detailed enough. Intro to predicates front page explanation provides a much friendly approach to the new paradigm.

### A New State-Minimized Transaction Model

By combining these state-minimized techniques with the UTXO model, Fuel offers a flexible transaction model that supports complex, multi-party transactions without the need for traditional smart contracts to access state. This approach also enhances light client verifiability, ensuring the security and integrity of transactions with minimal state footprint.

For a deeper analysis of how Fuel solves for Ethereum state bloat, read [Mastering the Final Boss in Blockchain Scalability: State Growth](https://fuel.mirror.xyz/OrgKx3EOE07YfPydGbQh1_R01lUqH0U0-uyrrEjpgrw) in the Fuel blog.

> ## Interoperability with Ethereum

Reviewer 2: Interoperability encompasses usability and application cohesion. I suggest including a section on Fuel's capability to accept various signatures, enabling wallets from different blockchain ecosystems to easily integrate with ours. From a user's perspective, there's no need to be concerned with the specific blockchain they are using; it simply works. 

I think its cool to include our Metamask demo: [Fuel Labs EVM Wallet Connector](https://github.com/FuelLabs/EVM-Wallet-Connector).

Author: Thank you, @calldelegation for the review and ideas. This is exactly the inspiration I was looking for on this last section. I'll get to enhancing this last section tomorrow!

> Fuel's commitment to interoperability with Ethereum is a cornerstone of its design, facilitating seamless integration between Fuel and the broader Ethereum ecosystem. This interoperability is not just a feature but a fundamental part of Fuel's architecture, enabling it to serve in a way that complements Ethereum's existing infrastructure.

Reviewer 2: ```suggestion
Fuel's commitment to interoperability with Ethereum is a cornerstone of its design, facilitating seamless integration between Fuel and the broader Ethereum ecosystem.
```

Reviewer 2: Kind of says the same thing

### Modular Architecture

> Fuel is designed with modularity at its core, allowing for easy interaction with various data availability and settlement layers. This flexibility ensures that Fuel can adapt to a range of scenarios, from operating alongside Ethereum as a Layer 2 solution to functioning independently or in conjunction with other systems. The modular nature of Fuel's architecture affirms its ability to provide high-performance execution without sacrificing the security and decentralized principles of the Ethereum network.

Reviewer 2: ```suggestion
Fuel is designed with modularity at its core, allowing for easy interaction with various data availability and settlement layers. This flexibility ensures that Fuel can adapt to a range of scenarios, from operating alongside Ethereum as a Layer 2 solution to functioning independently or remain agnostic to serve other blockchain architectures as the prime execution layer of choice. The modular nature of Fuel's architecture affirms its ability to provide high-performance execution without sacrificing the security and decentralized principles of the Ethereum network.
```

### Trust-Minimized Bridge

> The [Fuel native bridge](https://app.fuel.network/bridge) moves assets currently between Ethereum's Sepolia network and Fuel. It allows for seamless interoperability between Ethereum and Fuel. This bridging capability is essential for maintaining liquidity across the two ecosystems and ensures that users and developers can move assets freely.

Reviewer 2: ```suggestion
The [Fuel native bridge](https://app.fuel.network/bridge) moves assets currently between Ethereum and Fuel. This bridging capability is essential for maintaining liquidity across the two ecosystems and ensures that users and developers can move assets freely.
```

Reviewer 2: Testnets are always changing we can leave this part out as it will serve mainnet Ethereum in a couple months anyways.

The [Fuel native bridge](https://app.fuel.network/bridge) moves assets currently between Ethereum and Fuel. This bridging capability is essential for maintaining liquidity across the two ecosystems and ensures that users and developers can move assets freely.

### Application Cohesion

One of the key aspects of Fuel's interoperability with Ethereum is its focus on usability and application cohesion. This is demonstrated through Fuel's capability to accept various cryptographic signatures, which makes it easier for users to bring their own wallets to Fuel and for developers to build dApps on Fuel that can use most wallets.

#### EVM Wallet Connectors

Fuel's support for diverse signature standards ensures accessibility and placing user experience as a top priority. The [Fuel Wallet Connectors specifications](https://github.com/FuelLabs/fuels-wallet/wiki/Fuel-Wallet-Connectors) defines the flow and provides a system overview of how this is achieved, along with concise definitions of the methods in the [Fuel TypeScript SDK](https://docs.fuel.network/docs/fuels-ts/) and how to utilize them in the codebase of a dApp to enable generic EVM-compatiable wallet usage.

> The Fuel Metamask Demo, both the [sample code](https://github.com/FuelLabs/EVM-Wallet-Connector) and the [live demo](https://fuel-evm-wallet.vercel.app/) showcase how easily Ethereum users can connect and interact with Fuel using any EVM wallet. 

Reviewer 3: ```suggestion
The [Fuel Metamask Demo](https://fuel-evm-wallet.vercel.app/) and EVM Wallet Connector [repo](https://github.com/FuelLabs/EVM-Wallet-Connector) showcase how easily Ethereum users can connect and interact with Fuel using any EVM wallet. 
```

Author: Yes, agreed, that's a lot cleaner 💯
